nexus/src/app/external_ip.rs

```diff
@@ -144,30 +144,27 @@ impl super::Nexus {
         fip_selector: params::FloatingIpSelector,
         target: params::FloatingIpAttach,
     ) -> UpdateResult<views::FloatingIp> {
+        let fip_lookup =
+            self.floating_ip_lookup(opctx, fip_selector.clone())?;
+        let (.., authz_project, authz_fip) =
+            fip_lookup.lookup_for(authz::Action::Modify).await?;
+
         match target.kind {
             params::FloatingIpParentKind::Instance => {
                 // Handle the case where floating IP is specified by name (and
                 // therefore a project is given) but instance is specified by
                 // ID (and therefore the lookup doesn't want a project), as well
                 // as the converse: floating IP specified by ID (and no project
                 // given) but instance specified by name, and therefore needs
-                // a project. In the latter case, we have to fetch the floating
-                // IP by its ID in order to get the project to include with
-                // the instance.
-                let project =
-                    match (target.parent.clone(), fip_selector.clone().project)
-                    {
-                        (NameOrId::Id(_), _) => None,
-                        (NameOrId::Name(_), Some(p)) => Some(p),
-                        (NameOrId::Name(_), None) => {
-                            let fip_lookup = self.floating_ip_lookup(
-                                opctx,
-                                fip_selector.clone(),
-                            )?;
-                            let (.., fip) = fip_lookup.fetch().await?;
-                            Some(fip.project_id.into())
-                        }
-                    };
+                // a project. In the latter case, we need to place the floating
+                // IP's project ID into the instance selector.
+                let project = match (&target.parent, fip_selector.project) {
+                    (NameOrId::Id(_), _) => None,
+                    (NameOrId::Name(_), Some(p)) => Some(p),
+                    (NameOrId::Name(_), None) => {
+                        Some(authz_project.id().into())
+                    }
+                };
```

Comment:
Much cleaner! And actually... we don't even need to match on `fip_selector.project` anymore — we always know the project ID.

```rs
let instance_selector = params::InstanceSelector {
    project: match &target.parent {
        NameOrId::Id(_) => None,
        NameOrId::Name(_) => Some(authz_project.id().into()),
    },
    instance: target.parent,
};
```

Reply:
Fantastic; done.